rename TransactionError:ALREADY_IN_CHAIN

willcl-ark opened the pull request:
Closes: #19363

Renaming this error improves clarity around the returned error both internally and externally when a transactions' outputs are already found in the utxo set (`TransactionError::ALREADY_IN_CHAIN -> TransactionError::ALREADY_IN_UTXO_SET`)

theStack reviewed (approved):
Concept and code-review ACK aa422bcb898b7e3aa8dca912d92016bd22cf243d

I agree that the proposed names are more to the point and less confusing.

ismaelsadeeq commented on src/node/transaction.cpp:
```diff
@@ -22,7 +22,7 @@ static TransactionError HandleATMPError(const TxValidationState& state, std::str
     err_string_out = state.ToString();
     if (state.IsInvalid()) {
         if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
-            return TransactionError::MISSING_INPUTS;
```
Nice, I think we also rename the one in `TxValidationResult` enum.

g/`TX_MISSING_INPUTS`/`TX_MISSING_INPUT_OR_SPENT`/g
and update comments appropriately.

willcl-ark replied:
I did consider this, but it felt much less exposed to the user, only via `testmempoolaccept` AFAIK:

```cpp
    result_inner.pushKV("reject-reason", "missing-inputs");
```

Therefore as an "internal error", `TX_MISSING_INPUTS` (for any reason) it felt more accurate to me to leave as is? Very happy to be persuaded away from this viewpoint though :)

ismaelsadeeq replied:
> Therefore as an "internal error", TX_MISSING_INPUTS (for any reason) it felt more accurate to me to leave as is?

I was reviewing this based on the PR rationale in the OP. You stated that "a transaction's inputs are missing from the UTXO set, which could also be due to all inputs having already been spent (MISSING_INPUTS -> INPUTS_MISSING_OR_SPENT)." That's why I suggested that `TxValidationResult` one  should be changed just for uniformity. But Internally, you're right `TX_MISSING_INPUTS` inferred or spent.

Maybe just update the `testmempoolaccept` reject reason since it's user-facing?

But if you agree, you may have to update some functional tests that use it and maybe a release note since users heavily use the RPC.

This is a non blocking comment.

This PR is an improvement and will close the [issue](https://github.com/bitcoin/bitcoin/issues/19363)

ismaelsadeeq reviewed (approved):
Code Review ACK aa422bcb898b7e3aa8dca912d92016bd22cf243d

willcl-ark committed:
rpc: clarify ALREADY_IN_CHAIN rpc errors

When using `sendrawtransaction` the ALREADY_IN_CHAIN error help string
may be confusing.

Rename TransactionError::ALREADY_IN_CHAIN to
TransactionError::ALREADY_IN_UTXO_SET and update the rpc help string.

Remove backwards compatibility alias as no longer required.

willcl-ark commented:
Rebased on master to fix merge conflict.

Added an extra commit to cover `PSBTError::MISSING_INPUTS`.

ismaelsadeeq reviewed (approved):
re-ACK b2b4b932572c5bb1ffa3fc4f34e17130348fbc24

theStack reviewed (approved):
re-ACK b2b4b932572c5bb1ffa3fc4f34e17130348fbc24

ryanofsky commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -652,7 +652,7 @@ std::optional<PSBTError> LegacyScriptPubKeyMan::FillPSBT(PartiallySignedTransact
         if (input.non_witness_utxo) {
             if (txin.prevout.n >= input.non_witness_utxo->vout.size()) {
-                return PSBTError::MISSING_INPUTS;
+                return PSBTError::INPUTS_MISSING_OR_SPENT;
```
In commit "refactor: PSBTError::MISSING_INPUTS" (b2b4b932572c5bb1ffa3fc4f34e17130348fbc24)

Does the name INPUTS_MISSING_OR_SPENT actually make sense to describe these two errors from FillPSBT? I'm not sure, but it seems these errors would only happen if the PSBT data was not internally consistent, so maybe it would be more appropriate to call this error something like PSBTError::INPUTS_INVALID?

willcl-ark replied:
Yes I agree on re-reading this that this error does not quite correctly describe the specific check being failed here, which is that the outpoints vout is not higher than the number of vouts in the input, i.e. it is by definition invalid.

I have taken your suggestion on the wording in 77ef75d3ce4bff76a0db730b57896dcf7e8f3988

ryanofsky commented on src/common/messages.cpp:
```diff
@@ -78,8 +78,8 @@ bool FeeModeFromString(const std::string& mode_string, FeeEstimateMode& fee_esti
-        case PSBTError::MISSING_INPUTS:
-            return Untranslated("Inputs missing or spent");
+        case PSBTError::INPUTS_INVALID:
+            return Untranslated("Invalid inputs specified");
```
In commit "refactor: PSBTError::MISSING_INPUTS" (77ef75d3ce4bff76a0db730b57896dcf7e8f3988):

I think it would be a little better if this message said just "Invalid inputs" instead of "Invalid inputs specified" since the input indices are internal to the PSBT data structure, not really specified by the user.

Someone more familiar with PSBT code than me might be able to suggest a better error code and message than this to use, but at least with this change the error should be more accurate.

Note: a test currently fails because this string is changing:

```
File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/rpc_psbt.py", line 707, in run_test
  assert_raises_rpc_error(-25, 'Inputs missing or spent', self.nodes[0].walletprocesspsbt, 'cHNidP8BAJoCAAAAAkvEW8NnDtdNtDpsmze+Ht2LH35IJcKv00jKAlUs21RrAwAAAAD/////S8Rbw2cO1020OmybN74e3Ysffkglwq/TSMoCVSzbVGsBAAAAAP7///8CwLYClQAAAAAWABSNJKzjaUb3uOxixsvh1GGE3fW7zQD5ApUAAAAAFgAUKNw0x8HRctAgmvoevm4u1SbN7XIAAAAAAAEAnQIAAAACczMa321tVHuN4GKWKRncycI22aX3uXgwSFUKM2orjRsBAAAAAP7///9zMxrfbW1Ue43gYpYpGdzJwjbZpfe5eDBIVQozaiuNGwAAAAAA/v///wIA+QKVAAAAABl2qRT9zXUVA8Ls5iVqynLHe5/vSe1XyYisQM0ClQAAAAAWABRmWQUcjSjghQ8/uH4Bn/zkakwLtAAAAAAAAQEfQM0ClQAAAAAWABRmWQUcjSjghQ8/uH4Bn/zkakwLtAAAAA==')
```

https://cirrus-ci.com/task/5361689174999040?logs=ci#L3216

ryanofsky reviewed (approved):
Code review ACK 77ef75d3ce4bff76a0db730b57896dcf7e8f3988 if test failure is fixed (see comment below)

Note: I didn't totally follow the original issue or how these changes resolve it, but all the changes here seem reasonable regardless.

ryanofsky commented:
Code review ACK a05f4ca241045e8a4b295760590805f961c2e5e7.

Just changed psbt error string and fixed a test since the last review.

tdb3 commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -430,13 +430,13 @@ def sendrawtransaction_testmempoolaccept_tests(self):
             assert_equal(testres['allowed'], False)
             assert_equal(testres['reject-reason'], 'txn-already-known')
-            assert_raises_rpc_error(-27, 'Transaction already in block chain', node.sendrawtransaction, tx['hex'])
+            assert_raises_rpc_error(-27, 'Transaction outputs already in utxo set', node.sendrawtransaction, tx['hex'])
```
Although it's only the error message (and the number stays the same), I still think its appropriate add a release note for the change to the user.

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes
> Release notes should be written for any PR that:
> ...
> makes any other visible change to the end-user experience.

tdb3 reviewed (commented):
Approach ACK
Good work increasing clarity.  In the past I've had a head-scratching moment with some of these before I learned the nuances, so this should help people.

Left a comment about release notes.

Tested:
- Ran unit/functional tests locally (passed)
- Manually exercised the following scenario (regtest):
   - Created a transaction (and broadcasted).
   - Generated one block to confirm.
   - Attempted to call `sendrawtransaction` with the newly confirmed transaction.
   - Received `Transaction outputs already in UTXO set` as expected.
   - Performed `sendall` to spent all UTXOs.
   - Re-attempted to call `sendrawtransaction` with the the spent transaction.
   - Received `bad-txns-inputs-missingorspent`

maflcko reviewed (commented):
Left a question. Also, I don't think this is a "refactor"

ryanofsky reviewed (approved):
Agree on not calling the commits which change error messages refactoring. I do like first commit making error constant more consistent with error message, but agree it is not essential.

ismaelsadeeq reviewed (commented):
Looking good to me 🥇 

I will re-ACK after comments were addressed

willcl-ark committed:
doc: release note for 30212

tdb3 reviewed (approved):
ACK e9de0a76b99fd4708295e1178f6c079db92e7f36
Re-ran tests in https://github.com/bitcoin/bitcoin/pull/30212#pullrequestreview-2155193780 (successful)

ismaelsadeeq reviewed (commented):
ACK e9de0a76b99fd4708295e1178f6c079db92e7f36

ryanofsky reviewed (approved):
Code review ACK e9de0a76b99fd4708295e1178f6c079db92e7f36.

Since last review, change has been simplified, no longer renaming `PSBTError::MISSING_INPUTS` to `PSBTError::INPUTS_INVALID` and no longer renaming `TransactionError::MISSING_INPUTS` to `TransactionError::INPUTS_MISSING_OR_SPENT`

I did think both of those renames were improvements, but it does simplify the PR to drop them. Also it makes it easier to see how this PR fixes #19363.

Also release notes were added

Roasbeef commented:
Does `bitcoind` consider this to be a breaking change? The new error message may be "more accurate", but for clients that were matching on this error to figure out why a transaction was rejected, this breaks old behavior. 

Alternatively, the error constant could be renamed, with the error text/string left in place.
